### PR TITLE
[FeedManager][Storage] Updates `FeedManager` to deduct feeds by storage

### LINKED ISSUE #2590: [Feed][Storage] Ensure mapping of available feeds is accurate
The model checks the availability of feeds for multiple purposes - fulfilling ration requests, planning purchases are 2 of the big ones. Once #2575 is merged in, the mapping of harvested crops to storages will allow for the same crop grown in different fields to be stored in separate storages, each of which will map to a different rufas feed id.

## Expected Behavior

Accurately map what feeds are currently available for daily ration fulfillment. We need to ensure accuracy in assessing the current state of what crops are in which storages mapping to which feeds and the relative dates of those storage times. This is vital for feed purchase planning and accurate feed removal from storage to fulfill daily ration requests (done in the order of farmgrown feeds FIFO, then any purchased feed).

## Current Behavior

Currently if a user creates a crop config and rotation where the same crop (by name) is grown in separate fields, it can be mapped to separate feed storages which could each have separate feed ids. The way feed storage levels are assessed and used to plan purchasing and fulfill ration requests is by the crop name so we have the potential for mistakenly thinking there is more of a certain type of feed because there is a false positive erroneously based on the matching crop name.

## Possible Solution

While it may seem like a simple fix, the method for removing feeds from storage is complex/complicated and involves several factors which are heavily reliant on this previous method of harvested crop data which made sense when the crop data contained rufas id mapping info as well as storage info. that's no longer the case.

The process to address this should be redoing how feed storage inventory levels are assessed and implementing that properly in all the places where feed inventory levels are depended upon - for ration purchase planning, feed removal from inventory to fulfill daily ration requests, et al.

## Target Deadline

While this is technically an edge case, I spoke with @KevinP-B who said this was definitely something that would be useful down the line: the same crop likely needs to be able to be grown in multiple fields for effects of soil or management differences to be captured.

So I think the urgency of this is medium-high.

### PR DESCRIPTION
**Pull request by ew3361zh**
Corrects the edge case of fulfilling feed requests from farmgrown feed based on crop name rather than storage availability.

### Context

Issue(s) closed by this pull request: closes #2590

### What

Currently on dev there is a possibility that a feed being removed from storage to fulfill a ration request will mistake the amount of feed available because it's using crop name to search rather than storage feed availability. As the issue points out, because of how crops are now mapped to storages based on crop name AND field name and that the storage a crop is put in determines the rufas feed id, there's a possibility in searching for a feed based on crop name that you would get the wrong amount of availability.

This PR now changes the feed storage amount gathering mechanism to rely on actual storages with their singular feed ids and deducts the feeds from storages in the same FIFO farmgrown first, then purchased feed second.

### Why

This prevents the edge case of a feed request being fulfilled by the wrong feed because the same crop (by name) grown in different fields could map to different storages which would potentially result in it being stored as different rufas feed ids.

### How

The feed lookup process is now based on storages and their specific feed ids. Crop name is no longer taken into account when trying to fulfill a feed request or look at feed availability.

### Test plan

I think the main focus of testing should be looking at whether feeds are still being removed in the correct order which is farmgrown feed first, FIFO, then purchased feed with order being ignored.
A more in depth test would entail setting up a second instance of a crop that's already being planted and plant it in a separate field. Additionally, map that second crop to a different storage than the first crop is being stored in and map that storage to a separate rufas feed id. Then examine that the correct feed is being removed to fulfill the feed request and that the newly created storage with the different rufas feed id isn't also being removed. This isn't really possible with the way `_gather_available_feeds_by_id()` functions now anyways so I'm not sure how useful this would be to set up since it's a lot of setup.

### CODE REVIEW

File: RUFAS/biophysical/feed_storage/feed_manager.py [hunk 1]
```diff
@@ -561,55 +561,111 @@ def _deduct_feeds_from_inventory(
...
+        tuple[float, float]
+            The remaining amount of feed to deduct and the total amount deducted from storage.
+        """
+        deducted = 0.0
+        for storage in feed_storages:
+            if remaining <= 1e-3:
+                break
+            available = float(storage.dry_matter_mass)
+            if available <= 1e-3:
+                continue
```
Comment (allisterakun) [inline RUFAS/biophysical/feed_storage/feed_manager.py:662]: If "remaining" or "available" <= 1e-3, is it better if we manually set them to zero?

File: RUFAS/biophysical/feed_storage/feed_manager.py [hunk 2]
```diff
@@ -561,55 +561,111 @@ def _deduct_feeds_from_inventory(
...
+        total_purchased_deducted: dict[RUFAS_ID, float] = {}
+        total_farmgrown_deducted: dict[RUFAS_ID, float] = {}
+
+        for feed_id, amount_needed in feeds_to_deduct.items():
+            remaining = float(amount_needed)
+
+            remaining, farmgrown_deducted = self._deduct_from_storage(feed_id,
+                                                                      remaining,
+                                                                      farmgrown_by_id.get(feed_id, ()),
+                                                                      )
```
Comment (allisterakun) [inline RUFAS/biophysical/feed_storage/feed_manager.py]: Can we rename the "remaining" variable? It's a bit confusing with the amount of feed that remains in a storage.

**Review by allisterakun — APPROVED** [review]
Generally looks good, left some minor comments. Great job, Niko!

File: RUFAS/biophysical/feed_storage/feed_manager.py [hunk 3]
```diff
@@ -561,55 +561,111 @@ def _deduct_feeds_from_inventory(
...
+                self._om.add_error(
+                    "Feed Deduction Error",
+                    f"Not adequate feed to deduct remaining {remaining:.3f} kg DM of feed {feed_id}.",
+                    info_map,
+                )
+                raise ValueError(
+                    f"Not adequate feed to deduct remaining {remaining:.3f} kg DM of feed {feed_id}."
+                )
+
+        self._log_feed_deductions(total_purchased_deducted, total_farmgrown_deducted, simulation_day)
```
Comment (matthew7838) [inline RUFAS/biophysical/feed_storage/feed_manager.py]: At minimum, the `remaining` should be changed to something like `remaining_after_farm_grown`. Or can we even have the whole remaining tracking logic kept in this part of the code instead of passing it in and out of the function? I think it makes more sense to be here.

File: RUFAS/biophysical/feed_storage/feed_manager.py [hunk 4]
```diff
@@ -561,55 +561,111 @@ def _deduct_feeds_from_inventory(
...
+
+    def _log_feed_deductions(
+        self,
+        total_purchased: dict[RUFAS_ID, float],
+        total_farmgrown: dict[RUFAS_ID, float],
+        simulation_day: int,
+    ) -> None:
+        """Logs the total amounts of purchased and farmgrown feeds that were deducted.
```
Comment (matthew7838) [inline RUFAS/biophysical/feed_storage/feed_manager.py]:
```suggestion
        """
        Logs the total amounts of purchased and farmgrown feeds that were deducted.
```

File: RUFAS/biophysical/feed_storage/feed_manager.py [hunk 5]
```diff
@@ -561,55 +561,111 @@ def _deduct_feeds_from_inventory(
...
+
+        Parameters
+        ----------
+        total_purchased : dict[RUFAS_ID, float]
+            Mapping of RuFaS Feed IDs to the total amounts of that feed that were deducted from purchased feeds
+            (kg dry matter).
+        total_farmgrown : dict[RUFAS_ID, float]
+            Mapping of RuFaS Feed IDs to the total amounts of that feed that were deducted from farmgrown feeds
+            (kg dry matter).
```
Comment (matthew7838) [inline RUFAS/biophysical/feed_storage/feed_manager.py:618]: This is something that's kinda across the code base, but I think we should keep the units more consistent.
```suggestion
            Mapping of RuFaS Feed IDs to the total amounts of dry matter of that feed that were deducted from farmgrown feeds
            (kg).
```

**Review by matthew7838 — APPROVED** [review]
Great work, Niko. LGTM. Left a few comments.

File: RUFAS/biophysical/feed_storage/feed_manager.py (hunk 2 above)
Comment (ew3361zh) [reply RUFAS/biophysical/feed_storage/feed_manager.py]: What do you think about `remaining_amount_needed` which would reference the original `amount_needed` from the for-loop and distinguish it from something thinking it meant "remaining in storage"?

File: RUFAS/biophysical/feed_storage/feed_manager.py (hunk 3 above)
Comment (ew3361zh) [reply RUFAS/biophysical/feed_storage/feed_manager.py]: Something like this?

Screenshot 2025-10-09 at 2 29 07 PM

File: RUFAS/biophysical/feed_storage/feed_manager.py (hunk 5 above)
Comment (ew3361zh) [reply RUFAS/biophysical/feed_storage/feed_manager.py:618]: Yea, I thought we had a `kg DM` unit at one point - it looks like we still have a `kg CO2 / kg DM` unit but that seems problematic since the only reference to `kg CO2` is a separate unit which is `kg CO2-eq`. Maybe I will create an issue for someone to look into this. Thanks for highlighting!

**Comment by KFosterReed** [comment]
This is a big improvement but doesn't quite pass the test on the edge cases that Niko suggested (thanks for highlighting those!!). I used the attached input files and output filters to test 2 scenarios:

Test 1 - same crop (corn silage), two fields (field 1 and field 2), same feed ID (50), two different storages (corn_silage_1 and corn_silage_2)

Test 2 - same crop (corn silage), two fields (field 1 and field 2), two feed IDs (50, 51), two different storages (corn_silage_1 and corn_silage_2)

 Here is a short summary of my observations:

Test 1 - this seems to work as expected in that the total amount of cornsilage (feed ID 50) is reported in the FeedManager daily storage levels outputs that includes the amounts harvested from both fields. It is unclear whether the correct amounts are being stored in the Feed Storage Bunker outputs as they don't align with the amount isn the FeedManager outputs or the crop yeilds. There is also only one corn silage bunker instead of 2. I think this will be addressed in another PR but flagging it here. 

Test 2 - seems to mostly not work. The part that does work is that the correct amount of feed 50 is reported by FeedManager in the dialy storage levels. However, if run with the example diet that only uses feed 50, no feed  51 is reported by the FeedManager and as mentioned above, there is no way to tell where this specific feed is being stored since it doesn't have a separate feed storage. The bigger problem arises when I added feed 51 to the growing animal diet by replacing feed 50 in that diet with feed 51. When trying to test that, I got the following error: 

image



[example_freestall_dairy_metadata_test1.json](https://github.com/user-attachments/files/22954943/example_freestall_dairy_metadata_test1.json)
[example_freestall_dairy_metadata_test2.json](https://github.com/user-attachments/files/22954944/example_freestall_dairy_metadata_test2.json)
[example_small_field_corn_alf_silage_test.json](https://github.com/user-attachments/files/22954949/example_small_field_corn_alf_silage_test.json)
[example_small_field_corn_grain_alf_hay_test.json](https://github.com/user-attachments/files/22954950/example_small_field_corn_grain_alf_hay_test.json)
[example_feed_storage_configs_test1.json](https://github.com/user-attachments/files/22954951/example_feed_storage_configs_test1.json)
[example_feed_storage_configs_test2.json](https://github.com/user-attachments/files/22954952/example_feed_storage_configs_test2.json)
[example_freestall_feed_storages_test.json](https://github.com/user-attachments/files/22954953/example_freestall_feed_storages_test.json)
[example_sm_alf_corn_fertilizer_test.json](https://github.com/user-attachments/files/22954955/example_sm_alf_corn_fertilizer_test.json)
[report_feed_storage_summary.json](https://github.com/user-attachments/files/22954974/report_feed_storage_summary.json)

**Comment by ew3361zh** [comment]
Thank you, @KFosterReed, this is really helpful and thorough testing and I think will be helpful for how we continue to improve reporting for Feed Storage and FeedManager!

For test1, I'd like to pull in @matthew7838 and see if this is something that might be best addressed in his work on issue #2560.

For test2:
A couple things here: 
- We're currently not reporting anything for farmgrown feeds (within FeedManager) if the HarvestedCrop is mapped to a storage that has a feed-id that is not in the available feeds list. 
- Currently we add a warning for this:
Screenshot 2025-10-17 at 3 25 53 PM

- We do report the storage levels for a non-feed stored crop via the Storage class itself under `_record_storage_totals()` for each storage using that prefix: `self._prefix = f"Feed.{base_class_name}.{self.__class__.__name__}{self.storage_name}"`. But the `feed_id` is not currently in that report at all - I think this is because the storage reporting was set up prior to `feed_id` being a storage attribute. We could certainly add that to the outputs either here or in #2560.

- So we're recording the stored feeds that are in the user's specified diet via FeedManager's `report_stored_farmgrown_feeds()` and then ALL feeds in storages regardless of whether a stored harvestedcrop feed is part of the diet via Storage's `_record_storage_totals()`. That distinction seems somewhat helpful but could also see that being confusing.

- I was not able to replicate the error here - I tried switching the example Midwest feed json to use feed 51 instead of 50 and kept getting an Attribute error before being able to see the error you were mentioning. Could you show me what you updated to use feed 51 for this test example? Here's my error:
Screenshot 2025-10-20 at 9 26 40 AM

**Comment by KFosterReed** [comment]
Thanks for the explanations above @ew3361zh. So some good news is that I realized that I can actually see/collect outputs related to the individual feed storages via something like `Feed.Silage.Bunker.corn_silage_storage_1..` These outputs appear to be reported on a ration formulation interval so it is still challenging to know if they align with the amounts being harvested/deducted. Questions that I have related to the relationship between the amounts reported from the `Feed.Silage.Bunker.` outputs and the `report_stored_farmgrown_feeds` methods are:

- Are the the quantities in the `report_stored_farmgrown_feeds` methods being updated for the losses that occur during storage?
- What is the order of operations for updating the mass of feed in the Storage outputs? And why does this method never report the same mass as that in the `report_stored_farmgrown_feeds`  methods?

> I was not able to replicate the error here - I tried switching the example Midwest feed json to use feed 51 instead of 50 and kept getting an Attribute error before being able to see the error you were mentioning. Could you show me what you updated to use feed 51 for this test example? Here's my error:

I tested this again and still got the same/similar error :/ Here is the feed file where I changed the feed IDs so that there are two different corn silages being fed (50 and 51). Let me know what you come up with!

[example_Midwest_feed_test.json](https://github.com/user-attachments/files/23169316/example_Midwest_feed_test.json)

Also, here is an updated filter: 

[report_feed_storage_summary.json](https://github.com/user-attachments/files/23169747/report_feed_storage_summary.json)

**Comment by ew3361zh** [comment]
> * Are the the quantities in the `report_stored_farmgrown_feeds` methods being updated for the losses that occur during storage?

Yes! The `report_stored_farmgrown_feeds` method cycles through all `active_storages` which are storages that are specified by the user in their `feed_management` input (instantiated by `FeedManager._create_all_storages()` in the `FeedManager`'s init. This will take into account whatever storage losses occur to the stored crop through each storage's `process_degradations` cycle.

A couple of things to note here: 
1. We only `process_degradations` on the ration reformulation schedule (~30 days for the example simulations).
2. There are 2 separate calls to `report_stored_farmgrown_feeds` each with a separate suffix to try and simplify (really, haha) the feed storage level outputs: 1 is on a daily basis called from `SimulationEngine._daily_simulation()` and has the `"daily_storage_levels"` suffix added. The other is called on the ration reformulation schedule after `process_degradations` has been called and has the `"balance_storage_levels"` suffix added. This second is called from `SimulationEngine._formulate_ration()`

So that daily version won't have daily-updated degradations losses because we don't process degradations daily. The balance storage levels version will be reporting just on those days when we've processed the degradations.

> * What is the order of operations for updating the mass of feed in the Storage outputs? And why does this method never report the same mass as that in the report_stored_farmgrown_feeds methods?

I'm not sure I fully understand this question - do you mean what is the ooo between removing feed mass from storages to fulfill ration requests as well as degradations mass losses? As in specifically on a day when there is both a ration reformulation so degradations are processed as well as removal of a feed from storage to fulfill a ration?

In case that's it the daily order is:
1. harvest/store any crops
2. process degradations
3. fulfill (newly formulated) ration request

>  I tested this again and still got the same/similar error :/ Here is the feed file where I changed the feed IDs so that there are two different corn silages being fed (50 and 51). Let me know what you come up with!

Thank you, will let you know!

**Comment by ew3361zh** [comment]
> I tested this again and still got the same/similar error :/ Here is the feed file where I changed the feed IDs so that there are two different corn silages being fed (50 and 51). Let me know what you come up with!

Figured this out - there is plenty of feed 51 available at the time the deduction request is made. However, currently in `FeedManager` we're mapping `crop_name` to a specific `rufas_feed_id`. 

```python
self.crop_to_rufas_id: dict[str, RUFAS_ID] = {}
for storage in self.active_storages.values():
    if storage.rufas_feed_id in available_feed_ids:
        self.crop_to_rufas_id[str(storage.crop_name)] = storage.rufas_feed_id
```

This is used for checking what feeds we have available in storage (and also in the currently unused management for planning cycle purchases). In the current setup, if 2 crops have the same name but are in 2 different storages and map to 2 different rufas id feeds, one will overwrite the other.

When we then check to remove the available feeds in storage, the overwritten one does not show up as available so its available storage is not accounted for when we're looking at what storages we can deduct from.

For deduction purposes, we actually don't need the `crop_name` as requests are made by rufas ID and this could just be a list of `rufas_ids`. However, I believe @allisterakun has planning cycle purchases on his docket so maybe I shouldn't alter that `self.crop_to_rufas_id` map. I think I can instead just directly cross-reference the `available_feeds` list against current `active_storages` and then avoid this issue for now.

I ran a test with the scenario you provided and this was fixed under the change I made and will push to github shortly.

**Review by KFosterReed — APPROVED** [review]
This is a great improvement. It further enables users to specify which crop is being used for which feed and track the inventory including the use of multiple storages for the same feed! Great work NIko!
